lib/util/cloudfront_signed_url.ex

```diff
@@ -0,0 +1,86 @@
+# Copyright(c) 2015-2019 ACCESS CO., LTD. All rights reserved.
+
+use Croma
+
+defmodule Antikythera.CloudfrontSignedUrl do
+  @moduledoc """
+  This module provides `get_signed_url/4` to generate [a signed URL for CloudFront](https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/private-content-signed-urls.html)
+  Note that currently we support only a signed URL using a *canned policy*.
+  """
+
+  @doc """
+  Generates a signed URL to access a file via CloudFront.
+
+  ## Parameters
+
+  - `resource_url` (string): CloudFront URL used for accessing the file, including query string parameters, if any.
+  - `lifetime_in_seconds` (positive integer): Expiration time is determined as the sum of the current time (in seconds) and this value.
+  - `key_pair_id` (string): ID for an active CloudFront key pair used for generating the signature.
+  - `private_key` (string): RSA private key for the key pair specified by `key_pair_id`.
+
+  ## Return value
+
+  A generated signed URL (string).
+  """
+  defun get_signed_url(resource_url        :: v[String.t],
+                       lifetime_in_seconds :: v[pos_integer],
+                       key_pair_id         :: v[String.t],
+                       private_key         :: v[String.t]) :: String.t do
+    expires_in_seconds = System.system_time(:second) + lifetime_in_seconds
+    resource_url
+    |> URI.parse()
+    |> Map.update!(:query, fn query ->
+      case query do
+        nil   -> ""
```

Comment:
`URI.parse("http://example.com/?").query == ""`

Reply:
Thank you!
I changed the code to preserve original URL regardless of whether tailing `?`/`&` exists.
I also add some tests.
https://github.com/access-company/antikythera/pull/100/commits/6939a7d5d51baf25c812668d6a77dbf89319bd72